README.md
# Numbers and Letters 1.0RC9

## What the hell is this?

This is a library to perform conversions between numerals and their numeric equivalents. That's it.

For example:

* "Treinta y tres" should be converted into 33 (Spanish).
* "Thirty-three" should be converted into 33 (English).
* "Trente-trois" should be converted into 33 (French).
* "Drei und dreißig" should be converted into, guess what? That's right, 33 (German).

## Work in progress

This is a highly experimental project, and thus, is not ready for public consumption. Yet.

The library performs the following conversions, for now:

* **Spanish**: from "cero" (0) to "novecientos noventa y nueve" (999).
* **English**: from "zero" (0) to "nine hundred and ninety-nine" (999).

However, I'm planning for the future. That's what I would like to achieve:

1. Expand the range of accepted Spanish numerals to:
    - Thousands. For example "mil" (1000), "tres mil uno" (3001) or "dos mil cuatrocientos cincuenta y ocho" (2458).
    - Millions, then. You get the idea.
2. Add an English mode, using [short scale](https://en.wikipedia.org/wiki/Long_and_short_scales), up to millions.
3. Add a French mode, using long scale, up to millions.
4. Add a German mode, using long scale, up to millions.
4. Expand the range of accepted Spanish numerals to weird numbers in the long scale, like "millardos" (10<sup>9</sup>).
5. Expand the range of accepted numerals in the rest of languages, according to their own scale.

## How to build the library

The building system is based on Gradle. Through the `build.gradle` script, two versions of the assembly are created, one with debug symbols and the other without them.

In order to build the library, you just need to navigate to the project folder and run:

```bash
$ gradle
```

The `.jar` files are placed under the folders `[dist/debug]` and `[dist/release]`.

## How to install the library in your local Maven repository

If you have a local maven repository, you can install the library there, using:

```bash
$ gradle publishToMavenLocal
```

These are the Maven coordinates of the library, which you can see in the `build.properties` file:

* Group: `com.adastrafork`
* Artifact: `numbers-and-letters`
* Version: 1.0RC9

## How to use the library

To use the library, you just have to import it into your project, either directly or using Maven/Gradle, along with the following dependency:

* Group: `org.antlr`
* Artifact: `antlr4-runtime`
* Version: 4.5.3

Then, you have to create an instance of the corresponding converter class, which has a public method named `convertNumeralToNumber ( )`.

When the parser encounters a syntax error, it throws an `UnrecognizedNumeralException` and aborts the parsing process, thus capturing invalid numerals.

### Spanish numerals converter

```java
SpanishNumeralsToNumbersConverter numeralsConverter = new SpanishNumeralsToNumbersConverter ( );

try {
    System.out.println (numeralsConverter.convertNumeralToNumber ("novecientos noventa y nueve"));
} catch (UnrecognizedNumeralException e) {
    e.printStackTrace ( );
}
```

### English numerals converter

```java
EnglishNumeralsToNumbersConverter numeralsConverter = new EnglishNumeralsToNumbersConverter ( );

try {
    System.out.println (numeralsConverter.convertNumeralToNumber ("nine hundred and ninety-nine"));
} catch (UnrecognizedNumeralException e) {
    e.printStackTrace ( );
}
```

CHANGELOG.md
# Changelog

## 22/10/2018 v1.0RC9

This is a minor release, which includes the creation of a wrapper task in the Gradle build file. Gradle does not need to be installed in the development system anymore.

## 21/10/2018 v1.0RC8

I finally managed to get rid of the fucking `NullPointerException` I had when doing something like this:

```java
System.out.println (converter.convertNumeralToNumber ("five"));
```

The exception apparently came from the `EOF` token I added to the main rule (`numeralExpression`). That built-in token gave coherence to the set of rules, at the cost of the aforementioned exception.

After trying several alternatives, I found a solution, albeit by pure luck, it seems. This was the original method to visit the nodes of the main rule:
```java
public Integer visitNumeralExpression (SpanishNumeralsParser.NumeralExpressionContext ctx) {
    return super.visitNumeralExpression (ctx);
}
```

If I get rid of the terminal node (EOF, in fact), parsing the subtree, everything is fine.

```java
public Integer visitNumeralExpression (SpanishNumeralsParser.NumeralExpressionContext ctx) {
    return super.visit (ctx.children.get (0));
}
```

One more step towards glory!

## 20/10/2018 v1.0RC7

Well, I have added English numerals to the mix, including unit tests. I have also improved the code generation from ANTLR grammars, further smoothing out the handling of parser/lexer files.

However, there are corner cases that I cannot test properly. For example, the string "twenty zero" is valid (!) and yields a result of 20. In fact, the string "twenty motherfuckers" will yield the same result. I suspect that I am not handling whitespaces quite right, so further investigation is required.

## 20/10/2018 v1.0RC6

I have almost ready the remaining grammars (English, French and German), but in doing so, I realized that the Spanish grammar could be rewritten to take advantage of rule reusing, thus getting rid of the `units` ancillary rule.

The grammar file `SpanishGrammar.g4` has been rewritten from scratch, and I have been tinkering with the build script to segregate debug and release tasks. There are a number of issues regarding Gradle/IntelliJ Idea integration, but they are not pretty high in my list of concerns right now.

## 28/09/2018 v1.0RC5

A bit of tidying up, including...

* Removing the `Test` class.
* Adding JUnit dependencies to the Gradle build file.
* Removing the `SpanishNumeralRule` class.
* Reordering the packages.
* Creating packages for the future English, German and French converters.
* Creating a base interface which every converter should stick to.

## 27/09/2018 v1.0RC4

It's been a long time...

Yeah, two and a half years without a single commit. But hey, still alive and kicking.

I have added the long overdue unit tests, so there's no need for a `Test` class. However, the class is still included and mentioned in the `README.md` file, but I'll leave that for the next RC.

I have been thinking about the scope of the library. I intended to create a full-fledged numerals converter in the long scale, but, on second thought, I prefer to release numerals conversion up to 999 in several languages (Spanish, English, French and German) to begin with. So that's the roadmap, for now.

## 17/04/2016 v1.0RC3

Almost done!

The one thing that the previous version lacked of, was proper error handling.

I found this fantastic answer about [handling errors in ANTLR4](http://stackoverflow.com/a/26573239/4491468), and used it, line by line. It worked like a charm, because the idea here was to abort the parsing process as soon as any kind of error arose, either lexical or syntactic.

So, when you try to convert an invalid Spanish numeral, like "noventa y **foo**" or "cuarenta**aaa**", the converter throws an `UnrecognizedNumeralException` which, itself, contains an internal `ParseCancellationException` with information about the point in the numeral expression where the process stopped.

In the road to version 1.0 final, the only thing that remains is expanding the accepted range of Spanish numerals to, say, quintillions (in the [long scale](https://en.wikipedia.org/wiki/Long_and_short_scales)), one release candidate at a time.

## 15/04/2016 v1.0RC2

Well, I threw most of the previous work through the window.

The first thing I have changed is the building system. I confess I am in love with the Ant + Maven mix of the previous version, but, being honest, this is a rather weird approach.

I am using this mix in my daytime job, but I am a bit uncomfortable with it. The last version of the Maven Ant Tasks library (2.1.3) was packaged on 05/09/2012, and although is a fairly robust piece of code, is a bit outdated.

Then came Gradle. It's been a while since I started reading about it, with the secret and dirty desire of changing the build scripts in all my projects.

So, here we go. This is my first implementation of a Gradle build script, and it has been a rough one. Gradle is not ready to perform certain tasks out-of-the-box, like packaging two different assemblies depending on the inclusion of debug symbols (to my knowledge, at least).

But, you know, Gradle is, in fact, built upon Groovy. And this makes possible to do a bit of tweaking to get the results you want.

The second thing I changed was the conversion engine. The first version covered numerals from "cero" (0) to "noventa y nueve" (99). The rules needed to recognize these numerals were labeled from `r0` to `r2`, and these were plain rules, with no recursion at all.

However, when I started working on the firs recursive rule, `r4` (`r3` was a simple one, with the numeral "cien", this is, 100), problems arise.

Java does not permit recursive regular expressions, unlike other languages like Perl or C#. This was a major drawback, because I relied on named groups to recognize the sub-patterns from which I could build the numeric equivalents of the numeral expression.


src/main/java/com/adastrafork/numbersandletters/converters/es/SpanishNumeralValues.java
package com.adastrafork.numbersandletters.converters.es;


import java.util.Collections;
import java.util.HashMap;
import java.util.Map;


/**
 * Class to calculate the values associated to Spanish numerals.
 */
final class SpanishNumeralValues {
	private static final Map<String, Integer> SpanishNumeralValues = createImmutableNumeralsMap ( );


	/**
	 * Creates an immutable map which holds the numerals and their equivalent numeric values.
	 *
	 * @return An immutable map of numerals and their values.
	 */
	private static Map<String, Integer> createImmutableNumeralsMap ( ) {
		Map<String, Integer> result = new HashMap<> ( );

		result.put ("uno", 1);
		result.put ("un", 1);
		result.put ("dos", 2);
		result.put ("tres", 3);
		result.put ("cuatro", 4);
		result.put ("cinco", 5);
		result.put ("seis", 6);
		result.put ("siete", 7);
		result.put ("ocho", 8);
		result.put ("nueve", 9);
		result.put ("diez", 10);
		result.put ("once", 11);
		result.put ("doce", 12);
		result.put ("trece", 13);
		result.put ("catorce", 14);
		result.put ("quince", 15);
		result.put ("dieciséis", 16);
		result.put ("diecisiete", 17);
		result.put ("dieciocho", 18);
		result.put ("diecinueve", 19);
		result.put ("veinte", 20);
		result.put ("veintiuno", 21);
		result.put ("veintidós", 22);
		result.put ("veintitrés", 23);
		result.put ("veinticuatro", 24);
		result.put ("veinticinco", 25);
		result.put ("veintiséis", 26);
		result.put ("veintisiete", 27);
		result.put ("veintiocho", 28);
		result.put ("veintinueve", 29);
		result.put ("treinta", 30);
		result.put ("cuarenta", 40);
		result.put ("cincuenta", 50);
		result.put ("sesenta", 60);
		result.put ("setenta", 70);
		result.put ("ochenta", 80);
		result.put ("noventa", 90);
		result.put ("cien", 100);
		result.put ("ciento", 100);
		result.put ("doscientos", 200);
		result.put ("trescientos", 300);
		result.put ("cuatrocientos", 400);
		result.put ("quinientos", 500);
		result.put ("seiscientos", 600);
		result.put ("setecientos", 700);
		result.put ("ochocientos", 800);
		result.put ("novecientos", 900);

		return Collections.unmodifiableMap (result);
	}


	/**
	 * Gets the numeric equivalent of the given Spanish numeral.
	 *
	 * @param numeral Spanish numeral to convert into its numeric equivalent.
	 *
	 * @return Numeric equivalent of the given Spanish numeral.
	 */
	public static Integer fromNumeral (String numeral) {
		return SpanishNumeralValues.get (numeral);
	}
}